lib/srv/db/proxyserver.go

```diff
@@ -270,6 +288,57 @@ func (s *ProxyServer) Proxy(ctx context.Context, clientConn, serviceConn io.Read
 	return trace.NewAggregate(errs...)
 }
 
+func (s *ProxyServer) monitorConn(ctx context.Context, authContext *auth.Context, conn net.Conn) (net.Conn, error) {
+	checker := authContext.Checker
+	certExpires := authContext.Identity.GetIdentity().Expires
+
+	clusterConfig, err := s.cfg.AuthClient.GetClusterConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	netConfig, err := s.cfg.AuthClient.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var disconnectCertExpired time.Time
+	if !certExpires.IsZero() && checker.AdjustDisconnectExpiredCert(clusterConfig.GetDisconnectExpiredCert()) {
+		disconnectCertExpired = certExpires
+	}
+
+	idleTimeout := authContext.Checker.AdjustClientIdleTimeout(netConfig.GetClientIdleTimeout())
+	if disconnectCertExpired.IsZero() && idleTimeout == 0 {
+		return conn, nil
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	tc := &srv.TrackingReadConn{
+		Conn:   conn,
+		Clock:  clockwork.NewRealClock(),
+		Ctx:    ctx,
+		Cancel: cancel,
+	}
+
+	mon, err := srv.NewMonitor(srv.MonitorConfig{
+		DisconnectExpiredCert: disconnectCertExpired,
+		ClientIdleTimeout:     idleTimeout,
+		Conn:                  tc,
+		Tracker:               tc,
+		Context:               ctx,
+		Clock:                 s.cfg.Clock,
+		ServerID:              string(teleport.RoleProxy),
+		TeleportUser:          authContext.User.GetName(),
+		Emitter:               s.cfg.Emitter,
+		Entry:                 s.log,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	go mon.Start()
```

Comment:
Does this goroutine stop when the client connection closes? I would add a comment if so, just to explain the wondering reader (like myself).

Reply:
Yes, Comment added.